craft_parts/lifecycle_manager.py

```diff
@@ -65,7 +67,10 @@ def __init__(
         extra_build_packages: List[str] = None,
         **custom_args,  # custom passthrough args
     ):
-        # TODO: validate or slugify application name
+        if not application_name:
+            raise ValueError("application name cannot be empty")
+
+        application_name = re.sub("[^0-9A-Za-z_]+", "_", application_name)
```

Comment:
Nice!

Reply:
should we not match instead and raise? If not, maybe instead mention this normalization process in the doc string

Reply:
The transformation magic is cool but I agree that we shouldn't change what the application supplies. That could create problems at runtime.